### ISSUE TITLE
Deserializer Can't find a Class Program-ClassA (Should be Program.ClassA)

### ISSUE THREAD

**Issue by TheColonel2688** [wojtpl2/ExtendedXmlSerializer#232]
Running the below code results in the following error "ExtendedXmlSerializer.Core.Sprache.ParseException: 'An attempt was made to parse the identity 'clr-namespace:;assembly=StackTypeSerialization:Program-ClassA', but no type could be located with that name.'". using release 2.1.15 targeting .Net Core 2.1

```
using ExtendedXmlSerializer.Configuration;
using System.Collections.Generic;
using System.Xml;

public class Program
{
    public static void Main()
    {
        var outList = new List<ClassA>();

        var classB = new ClassB();
        var classC = new ClassC();

        outList.Add(new ClassA()
        {
            blah = "Test1",
            InterfaceConcreteTypeA = new ClassB() { TestInterfaceProperty = "Blah1", TestConcretePropertyB = "Blah1" },
            InterfaceConcreteTypeB = new ClassC() { TestInterfaceProperty = "Blah2", TestConcretePropertyC = "Blah2" }
        });
        outList.Add(new ClassA()
        {
            blah = "Test2",
            InterfaceConcreteTypeA = new ClassB(),
            InterfaceConcreteTypeB = new ClassC()
        });
        outList.Add(new ClassA()
        {
            blah = "Test3",
            InterfaceConcreteTypeA = new ClassB() { TestInterfaceProperty = "Blah3", TestConcretePropertyB = "Blah3" },
            InterfaceConcreteTypeB = new ClassC() { TestInterfaceProperty = "Blah3", TestConcretePropertyC = "Blah3" }
        });


        var serializer = new ConfigurationContainer().Create();

        XmlWriterSettings settings = new XmlWriterSettings
        {
            Indent = true,
            IndentChars = "\t"
        };

        using (XmlWriter writer = XmlWriter.Create("Test.xml", settings))
        {
            serializer.Serialize(writer, outList.ToArray());
        }

        List<ClassA> inList = new List<ClassA>();

        using (XmlReader reader = XmlReader.Create("Test.xml", new XmlReaderSettings { IgnoreWhitespace = false }))
        {
            inList = (List<ClassA>)serializer.Deserialize(reader);
        }

    }

    
    public class ClassA
    {        
        public string blah { get; set; }
        public InterfaceA InterfaceConcreteTypeA { get; set; }
        public InterfaceA InterfaceConcreteTypeB { get; set; }
    }

    public class ClassB : InterfaceA
    {
        public string TestInterfaceProperty { get; set; }
        public string TestConcretePropertyB { get; set; }
    }
    public class ClassC : InterfaceA
    {
        public string TestInterfaceProperty { get ; set; }
        public string TestConcretePropertyC { get; set; }
    }

    public interface InterfaceA
    {
        string TestInterfaceProperty { get; set; }
    }

}
```

![ctrls-sb-ben 2019-0130 11-35-05](https://user-images.githubusercontent.com/8753759/51996622-1dc0c880-2483-11e9-97af-c13b1626b13f.png)

**Comment by Mike-EEE** [comment]
Ah ok, I have some Real Life priorities of my own through today and tomorrow that might interfere here.  I will do my best to assist, regardless. :)

Are you getting the same error if you make the classes non-nested?  My fear here is that the issue has to do with a weird inheritance chain bug.  If the classes work outside of a container class, that might point to another issue altogether.

**Comment by TheColonel2688** [comment]
@Mike-EEE 

I was not 100% what you meant, but I tried the following. It is just creating a blank file, The first code sample seemed to correctly create the file it just wouldn't deserialize it.

```
        InterfaceA InterfaceConcreteTypeB = new ClassB() { TestConcretePropertyB = "BlahB", TestInterfaceProperty = "Interface Blah" };


        var serializer = new ConfigurationContainer().Create();

        XmlWriterSettings settings = new XmlWriterSettings
        {
            Indent = true,
            IndentChars = "\t"
        };
        
        using (XmlWriter writer = XmlWriter.Create("Test.xml", settings))
        {
            var blah = serializer.Serialize(settings, InterfaceConcreteTypeB);
        }
        
        //List<ClassA> inList = new List<ClassA>();

        using (XmlReader reader = XmlReader.Create("Test.xml", new XmlReaderSettings { IgnoreWhitespace = false }))
        {
            var inList = (InterfaceA)serializer.Deserialize(reader);
        }
```

**Comment by TheColonel2688** [comment]
So then I tried this.

This time I got an error messages that it could not find ClassB

```
InterfaceA InterfaceConcreteTypeB = new ClassB() { TestConcretePropertyB = "BlahB", TestInterfaceProperty = "Interface Blah" };


        var serializer = new ConfigurationContainer().Create();

        XmlWriterSettings settings = new XmlWriterSettings
        {
            Indent = true,
            IndentChars = "\t"
        };
        
        using (XmlWriter writer = XmlWriter.Create("Test.xml", settings))
        {
            serializer.Serialize(writer, InterfaceConcreteTypeB);
        }

        using (XmlReader reader = XmlReader.Create("Test.xml", new XmlReaderSettings { IgnoreWhitespace = false }))
        {
            //var inList = (List<InterfaceA>)serializer.Deserialize(reader);
            var theClass = (InterfaceA)serializer.Deserialize(reader);
        }
```

**Comment by TheColonel2688** [comment]
```        InterfaceA InterfaceConcreteTypeB = new ClassB() { TestConcretePropertyB = "BlahB", TestInterfaceProperty = "Interface Blah" };        var serializer = new ConfigurationContainer().Create();        XmlWriterSettings settings = new XmlWriterSettings        {            Indent = true,            IndentChars = "\t"        };                using (XmlWriter writer = XmlWriter.Create("Test.xml", settings))        {            var blah = serializer.Serialize(settings, InterfaceConcreteTypeB);        }                //List<ClassA> inList = new List<ClassA>();        using (XmlReader reader = XmlReader.Create("Test.xml", new XmlReaderSettings { IgnoreWhitespace = false }))        {            var inList = (InterfaceA)serializer.Deserialize(reader);        }```

**Comment by Mike-EEE** [comment]
Ahh... sorry for the confusion here @TheColonel2688.  What I meant by "container class" was `Progam`.  That is, you have all the classes listed as nested classes under the container class of `Program`.  My suggestion/thought was that there was something funky going on there.

However, that does not appear to be the case here.

I _finally_ got some time to look at this, and was able to check in a basic test that shows the object graph you initially provided being cycled successfully:

https://github.com/wojtpl2/ExtendedXmlSerializer/blob/850f31f85c2588f5f0881907bf353ba1e31dcf0e/test/ExtendedXmlSerializer.Tests/ReportedIssues/Issue232Tests.cs#L14-L54

So, it would seem that this is a problem that is lurking in the `XmlReader` which I will take a look into now.

I am not entirely confident that I will be able to look into this with the dedication it needs for today, however.  The good news is that tomorrow is completely free to work on this.

I would like to get you unblocked, of course.  Have you tried using the extension methods rather than using `XmlReader` and `XmlWriter`?

**Comment by Mike-EEE** [comment]
Welp, the mystery deepens, @TheColonel2688 	... I was able to copy/paste the code that you provided into a new .NET Core 2.1 csproj, and landed on the casting exception that we discussed:

![](https://i.imgur.com/WKXfeNN.png)

This is my csproj, FWIW:

```
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>netcoreapp2.1</TargetFramework>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="ExtendedXmlSerializer" Version="2.1.15" />
  </ItemGroup>

</Project>

```

**Comment by TheColonel2688** [comment]
I just tried this using the extension method that serializes to a string. I got the same result. (Note I have moved my tests classes out of the Program class in this example.

```
using ExtendedXmlSerializer.Configuration;
using ExtendedXmlSerializer.Core;
using ExtendedXmlSerializer.ReflectionModel;
using ExtendedXmlSerializer.ExtensionModel;
using ExtendedXmlSerializer.ContentModel;
using ExtendedXmlSerializer;
using System.Collections.Generic;
using System.Xml;
using ExtendedXmlSerializer.ExtensionModel.Xml;

public class Program
{
    public static void Main()
    {
        
        var outList = new List<ClassA>();

        var classB = new ClassB();
        var classC = new ClassC();

        outList.Add(new ClassA()
        {
            blah = "Test1",
            InterfaceConcreteTypeA = new ClassB() { TestInterfaceProperty = "Blah1", TestConcretePropertyB = "Blah1" },
            InterfaceConcreteTypeB = new ClassC() { TestInterfaceProperty = "Blah2", TestConcretePropertyC = "Blah2" }
        });
        outList.Add(new ClassA()
        {
            blah = "Test2",
            InterfaceConcreteTypeA = new ClassB(),
            InterfaceConcreteTypeB = new ClassC()
        });
        outList.Add(new ClassA()
        {
            blah = "Test3",
            InterfaceConcreteTypeA = new ClassB() { TestInterfaceProperty = "Blah3", TestConcretePropertyB = "Blah3" },
            InterfaceConcreteTypeB = new ClassC() { TestInterfaceProperty = "Blah3", TestConcretePropertyC = "Blah3" }
        });


        var serializer = new ConfigurationContainer().Create();

        XmlWriterSettings settings = new XmlWriterSettings
        {
            Indent = true,
            IndentChars = "\t"
        };

        string serializedAsString;

        serializedAsString = serializer.Serialize(settings, outList);

        var ReaderSettings = new XmlReaderSettings { IgnoreWhitespace = false };

        var blah = serializer.Deserialize<List<ClassA>>(ReaderSettings, serializedAsString);


    }
}

public class ClassA
{
    public string blah { get; set; }
    public InterfaceA InterfaceConcreteTypeA { get; set; }
    public InterfaceA InterfaceConcreteTypeB { get; set; }
}

public class ClassB : InterfaceA
{
    public string TestInterfaceProperty { get; set; }
    public string TestConcretePropertyB { get; set; }
}
public class ClassC : InterfaceA
{
    public string TestInterfaceProperty { get; set; }
    public string TestConcretePropertyC { get; set; }
}

public interface InterfaceA
{
    string TestInterfaceProperty { get; set; }
}
```

**Comment by TheColonel2688** [comment]
My .csproj is the same
```

<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>netcoreapp2.1</TargetFramework>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="ExtendedXmlSerializer" Version="2.1.15" />
  </ItemGroup>

</Project>
```

**Comment by TheColonel2688** [comment]
So I tried the same thing with .net 4.5. Same result.

Then I tried moving all of the classes to their own file. GUESS WHAT???!!!???! it worked....... this does not make ANY SENSE. whould should not matter what file they are in.

Then I thought, wait a minute there is no enclosing namespace {} in the main program.cs file.... So I added one and it worked........



![qwbz74a](https://user-images.githubusercontent.com/8753759/52209077-bd4dd480-2850-11e9-8e3c-511da99e8945.gif)

**Comment by Mike-EEE** [comment]
OMG that is hysterical!  So happy THAT was the problem!  I'll take those all day. 😂

Although, I probably shouldn't get too excited just yet.  Can you verify that everything works as expected now when you use the settings and also with the `XmlWriter.Create`/`XmlReader.Create`?

**Comment by TheColonel2688** [comment]
Yep it all works now even with the reader and writer.

**Comment by Mike-EEE** [comment]
Woohoo!  That makes me very happy as `XmlWriter` and `XmlReader` are not very good friends of mine. 😅  That was going to suck if I had to stick my hands in there again.

What a weird error though... it must have been resorting to or loading a previous compile, and/or secretly failing without telling you.

I am not going to worry about it, though. :P  Please let us know if you find any additional problems and we can take a look at it. 👍   Closing this one for now.

**Comment by TheColonel2688** [comment]
for posterity, this is complete working code

I am not sure what you mean by "secretly failing without telling me", it was definitely throwing an error on deserialize.

```

using ExtendedXmlSerializer.Configuration;
using System.Collections.Generic;
using System.Xml;
using ExtendedXmlSerializer.ExtensionModel.Xml;

namespace QuestionTypeSerialization
{

    public class Program
    {
        public static void Main()
        {

            var outList = new List<ClassA>();

            var classB = new ClassB();
            var classC = new ClassC();

            outList.Add(new ClassA()
            {
                blah = "Test1",
                InterfaceConcreteTypeA = new ClassB() { TestInterfaceProperty = "Blah1", TestConcretePropertyB = "Blah1" },
                InterfaceConcreteTypeB = new ClassC() { TestInterfaceProperty = "Blah2", TestConcretePropertyC = "Blah2" }
            });
            outList.Add(new ClassA()
            {
                blah = "Test2",
                InterfaceConcreteTypeA = new ClassB(),
                InterfaceConcreteTypeB = new ClassC()
            });
            outList.Add(new ClassA()
            {
                blah = "Test3",
                InterfaceConcreteTypeA = new ClassB() { TestInterfaceProperty = "Blah3", TestConcretePropertyB = "Blah3" },
                InterfaceConcreteTypeB = new ClassC() { TestInterfaceProperty = "Blah3", TestConcretePropertyC = "Blah3" }
            });


            var serializer = new ConfigurationContainer().Create();

            XmlWriterSettings settings = new XmlWriterSettings
            {
                Indent = true,
                IndentChars = "\t"
            };

            using (var xmlWriter = XmlWriter.Create("test.xml", settings))
            {
                serializer.Serialize(xmlWriter, outList);
            }


            var readerSettings = new XmlReaderSettings { IgnoreWhitespace = false };

            using (var xmlreader = XmlReader.Create("test.xml", readerSettings))
            {
                var blah = (List<ClassA>)serializer.Deserialize(xmlreader);
            }

        }
    }

    public class ClassA
    {
        public string blah { get; set; }
        public InterfaceA InterfaceConcreteTypeA { get; set; }
        public InterfaceA InterfaceConcreteTypeB { get; set; }
    }


    public class ClassB : InterfaceA
    {
        public string TestInterfaceProperty { get; set; }
        public string TestConcretePropertyB { get; set; }
    }
    public class ClassC : InterfaceA
    {
        public string TestInterfaceProperty { get; set; }
        public string TestConcretePropertyC { get; set; }
    }

    public interface InterfaceA
    {
        string TestInterfaceProperty { get; set; }
    }

}
```
